test/devise/migratable/model_test.rb

```diff
@@ -0,0 +1,170 @@
+require 'test_helper'
+require 'active_model'
+
+class MigratableTest < ActiveSupport::TestCase
+  include Support::Assertions
+  include Support::Factories
+
+  def unconfig_user_model
+    Class.new do
+      include ActiveModel::Serialization
+      include ActiveModel::Model
+      extend ActiveModel::Callbacks
+      include ActiveModel::Validations::Callbacks
+      extend Devise::Models
+      # in order to call update_attribute
+
+      define_model_callbacks :update, only: :after
+
+      devise :database_authenticatable, :migratable
+
+      attr_accessor :encrypted_password, :encrypted_password_migrate_to
+
+      def initialize(encrypted_password: '')
+        self.encrypted_password = encrypted_password
+        super
+      end
+
+      # skip active record layer
+      def update_attribute(_name, value)
+        self.encrypted_password_migrate_to = value
+      end
+
+      # prevent moving salt everytime
+      def self.password_salt
+        'here we go'
+      end
+
+      def attributes
+        { 'encrypted_password' => nil, 'encrypted_password_migrate_to' => nil }
+      end
+    end
+  end
+
+  def configed_user_model
+    Class.new(unconfig_user_model) do
+      devise :database_authenticatable, :migratable, encryptor: :pbkdf2_sha512
+    end
+  end
+
+  def random_password_salt_user_model
+    Class.new(configed_user_model) do
+      def self.password_salt
+        Devise.friendly_token[0, 20]
+      end
+    end
+  end
+
+  def user_model_with_additional_except_list
+    Class.new(configed_user_model) do
+      attr_accessor :phone
+
+      def serializable_hash(options = {})
+        options = options.try(:dup) || {}
+        options[:except] ||= %i[phone]
+        super options
+      end
+
+      def attributes
+        { 'encrypted_password' => nil, 'encrypted_password_migrate_to' => nil, 'phone' => nil }
+      end
+    end
+  end
+
+  def feature_class(active: true)
+    Class.new do
+      define_singleton_method :active? do |_name, _model|
+        active
+      end
+    end
+  end
+
+  def configed_user_model_with_feature(enabled: true)
+    enabled_feature_class = feature_class(active: enabled)
+    Class.new(unconfig_user_model) do
+      devise :database_authenticatable,
+             :migratable,
+             encryptor: :pbkdf2_sha512,
+             feature_class: enabled_feature_class,
+             feature_name: :does_not_matter
+    end
+  end
+
+  test 'should generate salt while setting password' do
+    user = configed_user_model.new
+    user.password = 'password'
+
+    new_encrypted_pass = '$pbkdf2-sha512$1$Hr4A4Ag$CMvUhq6jnDgfYLtLGYlI30t5kDl/enxTd0ATWSuK2NmekdY9osewu3b4KEaAhygMipdeE.9fL0Ur56bnDpQkTQ'
```

Comment:
I was gonna mention in the other file that I would prefer the plain encrypted string than

```ruby
hashed_password = Devise::Migratable::Encryptors::Pbkdf2Sha512.digest(..)
```
Glad that this test exists here.